### PR TITLE
Release/0.7.0

### PR DESCRIPTION
**Pull request by oguzhanunlu**
For #106 , following statements are to support storage integration:

- COPY INTO TABLE
- CREATE STAGE

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -18,7 +18,7 @@ Assuming git and [SBT][sbt] installed:
 ```bash
 $ git clone https://github.com/snowplow-incubator/snowplow-snowflake-loader.git
 $ cd snowplow-snowflake-loader
-$ sbt test
+$ sbt clean startDynamodbLocal test dynamodbLocalTestCleanup stopDynamodbLocal
```
Comment (chuwy) [inline README.md]: Would it fail without DynamoDB? Just don't want to pollute the quick start guide.

File: loader/src/main/scala/com.snowplowanalytics.snowflake.loader/PasswordService.scala [hunk 2]
```diff
@@ -33,23 +33,29 @@ object PasswordService {
...
+   * Get credentials for `setup` subcommand
+   * Only IAM keys and storage integration are supported
+   */
+  def getSetupCredentials(authMethod: Config.AuthMethod): Option[Common.Auth] =
     authMethod match {
       case Config.AuthMethod.CredentialsAuth(accessKeyId, secretAccessKey) =>
-        Some(Common.AwsCreds(accessKeyId, secretAccessKey, None))
+        Some(Common.AwsKeys(accessKeyId, secretAccessKey, None))
+      case Config.AuthMethod.StorageIntegration(name) => Some(Common.StorageIntegration(name))
       case _ => None
```
Comment (chuwy) [inline loader/src/main/scala/com.snowplowanalytics.snowflake.loader/PasswordService.scala]: Can you add please all other options explicitly?

File: loader/src/main/scala/com.snowplowanalytics.snowflake.loader/ast/Common.scala [hunk 3]
```diff
@@ -13,5 +13,31 @@
 package com.snowplowanalytics.snowflake.loader.ast
 
 object Common {
```
Comment (chuwy) [inline loader/src/main/scala/com.snowplowanalytics.snowflake.loader/ast/Common.scala]: I'd replace the whole `Common` module with following:

```scala
sealed trait Auth

object Auth {
  final case class StorageIntegration(name: String) extends Auth
  final case class AwsKeys(awsAccessKeyId: String, awsSecretKey: String, sessionToken: Option[String]) extends AwsCreds
...
}
```

File: loader/src/main/scala/com.snowplowanalytics.snowflake.loader/ast/Common.scala [hunk 4]
```diff
@@ -13,5 +13,31 @@
...
+   * Represents a Storage Integration entity in Snowflake
+   * The recommended way to setup authentication
+   * @param name Name of the storage integration
+   */
+  final case class StorageIntegration(name: String) extends Auth
+
+  /**
+   * Represents all AWS-based authentication options supported by Snowflake
+   */
+  sealed trait AwsCreds extends Auth
```
Comment (chuwy) [inline loader/src/main/scala/com.snowplowanalytics.snowflake.loader/ast/Common.scala]: Why do we need this second-level? `AwsRole` doesn't feel "credentials" to me, all three are equal authentication mechanisms to me.

File: loader/src/main/scala/com.snowplowanalytics.snowflake.loader/ast/Statement.scala [hunk 5]
```diff
@@ -55,15 +55,17 @@ object Statement {
...
-        case Some(c) => s" CREDENTIALS = (AWS_KEY_ID = '${c.awsAccessKeyId}' AWS_SECRET_KEY = '${c.awsSecretKey}')"
-        case None => ""  // Expect credentials are available in stage
+      val auth = ddl.credentials.fold(""){
+        case Common.AwsKeys(accessKey, secretKey, token) =>
+          if (token.isDefined) {
+            System.err.println("AWS_TOKEN (temporary credentials) must never be used for stages. Skipping credentials")
+            ""
+          } else s" CREDENTIALS = ( AWS_KEY_ID = '$accessKey' AWS_SECRET_KEY = '$secretKey' )"
+        case Common.AwsRole(arn) => s" CREDENTIALS = ( AWS_ROLE = '$arn' )"
+        case Common.StorageIntegration(_) => s"" // external stage is to provide authentication
```
Comment (chuwy) [inline loader/src/main/scala/com.snowplowanalytics.snowflake.loader/ast/Statement.scala]: Does it mean that we don't care about name in the end? Or rather do, but only on setup stage.

File: loader/src/main/scala/com.snowplowanalytics.snowflake.loader/connection/Jdbc.scala [hunk 6]
```diff
@@ -31,15 +31,18 @@ object Jdbc {
...
-      // US West is default: https://docs.snowflake.net/manuals/user-guide/jdbc-configure.html#jdbc-driver-connection-string
+      // US West (Oregon), us-west-2, is default: https://docs.snowflake.com/en/user-guide/jdbc-configure.html#connection-parameters
       val host = config.jdbcHost match {
         case Some(overrideHost) => overrideHost
         case None =>
-          if (config.snowflakeRegion == "us-west-1")
+          if (config.snowflakeRegion == "us-west-2")
             s"${config.account}.snowflakecomputing.com"
-          else
+          else if (List("us-east-1", "eu-west-1", "eu-central-1", "ap-southeast-1", "ap-southeast-2").contains(config.snowflakeRegion))
```
Comment (chuwy) [inline loader/src/main/scala/com.snowplowanalytics.snowflake.loader/connection/Jdbc.scala]: Let's define it as outside constant.

File: project/build.properties [hunk 7]
```diff
@@ -1 +1 @@
-sbt.version=1.2.6
+sbt.version=1.2.8
```
Comment (chuwy) [inline project/build.properties]: Why not 1.3.13?

File: README.md (hunk 1 above)
Comment (chuwy) [reply README.md]: Honestly I'd change it to assembly as well. I imagine users would like to assembly it more often.

File: loader/src/main/scala/com.snowplowanalytics.snowflake.loader/ast/Common.scala (hunk 4 above)
Comment (oguzhanunlu) [reply loader/src/main/scala/com.snowplowanalytics.snowflake.loader/ast/Common.scala]: I don't have strong opinion here, I'll update accordingly

File: loader/src/main/scala/com.snowplowanalytics.snowflake.loader/PasswordService.scala (hunk 2 above)
Comment (oguzhanunlu) [reply loader/src/main/scala/com.snowplowanalytics.snowflake.loader/PasswordService.scala]: sure

File: README.md (hunk 1 above)
Comment (oguzhanunlu) [reply README.md]: > Would it fail without DynamoDB?

yes

> I'd change it to assembly as well. 

I'll update accordingly

File: loader/src/main/scala/com.snowplowanalytics.snowflake.loader/ast/Statement.scala (hunk 5 above)
Comment (oguzhanunlu) [reply loader/src/main/scala/com.snowplowanalytics.snowflake.loader/ast/Statement.scala]: > Or rather do, but only on setup stage.

yes

File: project/build.properties (hunk 7 above)
Comment (oguzhanunlu) [reply project/build.properties]: no reason, will do so

File: loader/src/main/scala/com.snowplowanalytics.snowflake.loader/Loader.scala [hunk 8]
```diff
@@ -127,32 +127,43 @@ object Loader {
...
+      stage.get("url").fold(s"No url info is available for stage [${conf.stage}]".invalidNel[Unit]){ sfStageUrl =>
+        if (sfStageUrl == conf.stageUrl.toString) ().validNel[String]
+        else (s"Stage [${conf.stage}] is configured to use url [$sfStageUrl] which does not match " +
+          s"stageUrl [${conf.stageUrl}] provided in loader configuration").invalidNel[Unit]
+      }
+
+    val storageIntegrationCheck: (Map[String, Object], AuthMethod) => ValidatedNel[String, Unit] = (stage, auth) =>
+      auth match {
+        case AuthMethod.StorageIntegration(intName) =>
+          stage.get("storage_integration").fold(s"No storage integration is found for stage [${conf.stage}]".invalidNel[Unit]){ sfIntName =>
```
Comment (chuwy) [inline loader/src/main/scala/com.snowplowanalytics.snowflake.loader/Loader.scala:148]: I'd much prefer just plain pattern match here, but it's very personal, I guess.

File: loader/src/main/scala/com.snowplowanalytics.snowflake.loader/Loader.scala [hunk 9]
```diff
@@ -127,32 +127,43 @@ object Loader {
...
+            if (sfIntName == intName.toUpperCase) ().validNel[String]
+            else (s"Stage [${conf.stage}] is configured to use integration [$sfIntName] which does not match " +
```
Comment (chuwy) [inline loader/src/main/scala/com.snowplowanalytics.snowflake.loader/Loader.scala:150]: As of `("smth" + "else")` you can use multiline strings.

**Review by chuwy — APPROVED** [review]
👍

**Pull request merged**